src/hermes/model/context.py

```diff
@@ -0,0 +1,215 @@
+    parent context.
+    """
+
+    def __init__(self, base: HermesContext, ep: EntryPoint):
+        """
+        Initialize a new harvesting context.
+
+        :param base: The base HermesContext that should receive the results of the harvesting.
+        :param ep: The entry point that implements the harvester using this context.
+        """
+
+        super().__init__()
+
+        self._base = base
+        self._ep = ep
+        self._log = logging.getLogger(f'harvest.{self._ep.name}')
+
+    def load_cache(self):
+        """
+        Load the cached data from the :py:attr:`HermesContext.hermes_dir`.
+        """
+
+        data_file = self._base.get_cache('harvest', self._ep.name)
+        if data_file.is_file():
+            self._log.debug("Loading cache from %s...", data_file)
+            self._data = json.load(data_file.open('r'))
+
+    def store_cache(self):
+        """
+        Store the collected data to the :py:attr:`HermesContext.hermes_dir`.
+        """
+
+        data_file = self.get_cache('harvest', self._ep.name, create=True)
+        self._log.debug("Writing cache to %s...", data_file)
+        json.dump(self._data, data_file.open('w'))
+
+    def __enter__(self):
+        self.load_cache()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.store_cache()
+        if exc_type is not None and issubclass(exc_type, HermesValidationError):
+            exc = traceback.TracebackException(exc_type, exc_val, exc_tb)
+            self._base.error(self._ep, exc)
+            return True
+
+    def update(self, _key: str, _value: t.Any, **kwargs: t.Any):
+        """
+        The updates are added to a list of values.
+        A value is only replaced if the `_key` and all `kwargs` match.
+
+        .. code:: python
+
+            # 'value 2' will be added (twice)
+            ctx.update('key', 'value 1', spam='eggs')
+            ctx.update('key', 'value 2', foo='bar')
+            ctx.update('key', 'value 2', foo='bar', spam='eggs')
+
+            # 'value 2' will replace 'value 1'
+            ctx.update('key', 'value 1', spam='eggs')
+            ctx.update('key', 'value 2', spam='eggs')
+
+        This way, the harvester can fully specify the source and only override values that are from the same origin
+        (e.g., if the data changed between two runs).
+
+        See :py:meth:`HermesContext.update` for more information.
+        """
+
+        if _key not in self._data:
+            self._data[_key] = []
+
+        for entry in self._data[_key]:
+            if entry[1] == kwargs:
+                self._log.debug("Update %s: %s -> %s (%s)", _key, entry[0], _value, entry[1])
+                entry[0] = _value
+                break
+        else:
+            self._data[_key].append([_value, kwargs])
+
+    def _update_key_from(self, _key: str, _value: t.Any, **kwargs):
+        if isinstance(_value, dict):
+            for key, value in _value.items():
```

Comment:
Should we add a similar behaviour for arrays?

Reply:
I think dictionary is fine for now.